File: articles/cell-theory-basic-unit-of-life/r/growth_models.R
# Cell-growth model workflows in R.

counts_path <- file.path("data", "cell_counts.csv")

if (!file.exists(counts_path)) {
  counts_path <- file.path("..", "data", "cell_counts.csv")
}

counts <- read.csv(counts_path)

fit_condition <- function(df) {
  fit <- lm(log(cells) ~ time_h, data = df)

  r_est <- coef(fit)[["time_h"]]
  n0_est <- exp(coef(fit)[["(Intercept)"]])
  td_est <- log(2) / r_est

  data.frame(
    condition = unique(df$condition),
    growth_rate_per_h = r_est,
    initial_count = n0_est,
    doubling_time_h = td_est,
    r_squared_log_space = summary(fit)$r.squared
  )
}

fit_df <- do.call(
  rbind,
  lapply(split(counts, counts$condition), fit_condition)
)

print(round(fit_df, 5))

File: articles/cell-theory-basic-unit-of-life/python/growth_models.py
"""
Cell-growth model workflows.

Run:
    python python/growth_models.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd

from cell_theory_core import fit_exponential_growth, logistic_growth


ARTICLE_DIR = Path(__file__).resolve().parents[1]
COUNTS_PATH = ARTICLE_DIR / "data" / "cell_counts.csv"
LOGISTIC_PATH = ARTICLE_DIR / "data" / "logistic_scenarios.csv"


def main() -> None:
    """Fit exponential growth and simulate logistic scenarios."""

    counts = pd.read_csv(COUNTS_PATH)

    fit_rows = []

    for condition, group in counts.groupby("condition"):
        fit = fit_exponential_growth(
            group["time_h"].to_numpy(dtype=float),
            group["cells"].to_numpy(dtype=float),
        )

        fit_rows.append(
            {
                "condition": condition,
                "growth_rate_per_h": fit.growth_rate_per_h,
                "initial_count": fit.initial_count,
                "doubling_time_h": fit.doubling_time_h,
                "r_squared_log_space": fit.r_squared_log_space,
            }
        )

    scenarios = pd.read_csv(LOGISTIC_PATH)
    logistic_rows = []

    for _, scenario in scenarios.iterrows():
        time = np.arange(0, scenario["time_end"] + scenario["dt"], scenario["dt"])

        trajectory = logistic_growth(
            time_h=time,
            initial_count=scenario["initial_count"],
            growth_rate=scenario["growth_rate"],
            carrying_capacity=scenario["carrying_capacity"],
        )

        logistic_rows.append(
            {
                "scenario": scenario["scenario"],
                "final_cell_count": trajectory[-1],
                "fraction_of_capacity": trajectory[-1] / scenario["carrying_capacity"],
                "initial_doubling_time_h": np.log(2) / scenario["growth_rate"],
            }
        )

    print(pd.DataFrame(fit_rows).round(5).to_string(index=False))
    print(pd.DataFrame(logistic_rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
